File: src/VirtoCommerce.PageBuilderModule.Web/Apps/page-builder-designer/src/app/modules/editor/models/page-history.model.ts
/**
 * A version of the open page: one commit that touched its file, wherever in the content repository it
 * lives. Only stores whose pages are kept in git have these — the server offers the `history` descriptor
 * for those and withholds it for the rest, which is what hides the panel entirely.
 */
export interface PageVersion {
    sha: string;
    shortSha: string;
    date?: string;
    message?: string;
    author?: {
        name?: string;
        email?: string;
        /** GitHub account the author's email resolves to, when it resolves to one. */
        login?: string;
    };
    /** Platform login recorded by the server; absent for edits made outside the builder. */
    vcUser?: string;
    /** Branches the commit is reachable from — a version can sit on several at once. */
    branches: string[];
    /** The production branch has this commit: it is what visitors see. */
    published: boolean;
    /** This version is on my own work branch. */
    mine: boolean;
    /** A commit that changed far more than this page — an import or a bulk reformat. */
    bulk: boolean;
    changedFiles?: number;
}

export interface PageHistory {
    versions: PageVersion[];
    /** More branches exist than the server looked at, so the list may be missing unpublished versions. */
    truncated: boolean;
    endCursor?: string;
    /** Unpublished versions that are neither mine nor bulk — the number on the toolbar badge. */
    otherDraftCount: number;
    myDraftBranch?: string;
}

/** Per-page history as the store holds it, including what is going on with it right now. */
export interface PageHistoryState extends PageHistory {
    isLoading: boolean;
    error?: string;
    /** Sha being restored, so the row that was clicked can say so. */
    restoring?: string;
}

/**
 * Consecutive versions by the same author on the same branch, folded into one row.
 *
 * Saving is a commit, so an afternoon of editing is a stack of near-identical entries — twelve of them
 * within three hours on one page in the content repository. The newest carries the group; the rest are
 * there to be expanded.
 */
export interface PageVersionGroup {
    version: PageVersion;
    older: PageVersion[];
}

File: src/VirtoCommerce.PageBuilderModule.Web/Models/PageHistoryModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using VirtoCommerce.PageBuilderModule.Core.GitContent;

namespace VirtoCommerce.PageBuilderModule.Web.Models
{
    /// <summary>
    /// A page's versions as the builder sees them: the repository's answer plus the one thing only the
    /// server can add — which of them are the current editor's own.
    /// </summary>
    public class PageHistoryModel
    {
        public IList<PageVersionModel> Versions { get; set; } = [];

        /// <summary>
        /// The repository has more branches than this answer looked at, so there may be unpublished
        /// versions it does not list. Passed on rather than hidden: a list that is quietly incomplete
        /// reads as "there is nothing else".
        /// </summary>
        public bool Truncated { get; set; }

        /// <summary>Pass back as <c>after</c> to continue the branch scan.</summary>
        public string EndCursor { get; set; }

        /// <summary>
        /// Unpublished versions that are not this editor's own and not bulk changes — what the toolbar
        /// counts on its badge. Their own draft is what the designer already shows them, and a
        /// repository-wide import is nobody's draft of this page, so neither belongs in a number that
        /// means "someone else has unpublished work here".
        /// </summary>
        public int OtherDraftCount { get; set; }

        /// <summary>The branch this editor's own draft of the page lives on, whether or not it exists yet.</summary>
        public string MyDraftBranch { get; set; }

        public static PageHistoryModel Empty => new();

        public static PageHistoryModel From(GitPageHistory history, string myBranch)
        {
            ArgumentNullException.ThrowIfNull(history);

            var versions = history.Versions
                .Select(version => PageVersionModel.From(version, myBranch))
                .ToList();

            return new PageHistoryModel
            {
                Versions = versions,
                Truncated = history.Truncated,
                EndCursor = history.EndCursor,
                OtherDraftCount = versions.Count(version => !version.Published && !version.Mine && !version.Bulk),
                MyDraftBranch = myBranch,
            };
        }
    }

    public class PageVersionModel
    {
        public string Sha { get; set; }

        public string ShortSha { get; set; }

        public DateTime? Date { get; set; }

        public string Message { get; set; }

        public PageVersionAuthorModel Author { get; set; }

        /// <summary>Platform login the server recorded in the commit; absent for edits made outside the module.</summary>
        public string VcUser { get; set; }

        public IList<string> Branches { get; set; } = [];

        public bool Published { get; set; }

        /// <summary>
        /// This version is on the current editor's own work branch. Computed from the branch name, which
        /// is derived from their login — so it is exact for anything the builder committed, and false for
        /// an edit they made outside it under a different git identity (see the identity increment).
        /// </summary>
        public bool Mine { get; set; }

        /// <summary>A commit that changed far more than this page — an import or a bulk reformat.</summary>
        public bool Bulk { get; set; }

        public int? ChangedFiles { get; set; }

        public static PageVersionModel From(GitPageVersion version, string myBranch) => new()
        {
            Sha = version.Sha,
            ShortSha = version.ShortSha,
            Date = version.Date,
            Message = version.Message,
            Author = new PageVersionAuthorModel
            {
                Name = version.AuthorName,
                Email = version.AuthorEmail,
                Login = version.AuthorLogin,
            },
            VcUser = version.VcUser,
            Branches = version.Branches,
            Published = version.Published,
            Mine = !string.IsNullOrEmpty(myBranch) && version.Branches.Contains(myBranch, StringComparer.Ordinal),
            Bulk = version.Bulk,
            ChangedFiles = version.ChangedFiles,
        };
    }

    public class PageVersionAuthorModel
    {
        public string Name { get; set; }

        public string Email { get; set; }

        /// <summary>GitHub account the author email resolves to, when it resolves to one.</summary>
        public string Login { get; set; }
    }
}
